File: src/engine/adapters/s3/caps.ts
import type { Caps } from '../../../shared/caps';

// §5.1's s3 row: reuses the keyvalue shape (page.ts's own doc comment on KeyValuePage explains
// why — a single object's metadata+body is exactly a flat field/value listing, same as a redis
// hash). Read-only browsing only in this phase (P17): no insert/update/delete, no console.
export const s3Caps: Caps = {
  tabular: false,
  documents: false,
  keyValue: true,
  stream: false,
  defaultPageKind: 'keyvalue',
  sql: false,
  // P23 D11: stays false for now, as a named follow-up rather than a permanent no — an *object*
  // already shows its full metadata in the keyvalue view it opens into (P17), so only a *bucket*
  // has anything new, and a bucket's properties are five separate SDK calls each of which a
  // single-bucket IAM policy routinely denies (catalog.ts's own comment on that policy shape).
  // Doing it properly means per-call degradation to a `notes` line per denial — its own piece of
  // work, not something this phase's Kafka/SQS pattern can just extend.
  definition: false,
  // describe() throws E_UNSUPPORTED (s3/index.ts). definition is already false above, so this is
  // a coincidence of two unrelated flags (P31 F5), not something describe: false relies on.
  describe: false,
  projection: false,
  serverFilter: false,
  // count() (s3/read.ts's countObject) answers a single object's own field count via HeadObject,
  // which is always exact — the same per-item-exact resolution redis/caps.ts makes for its own
  // per-key counts, not the db-wide "how many keys total" question ListObjectsV2 would need to
  // answer approximately (and which this phase's read-only object browsing never asks).
  exactCount: true,
  pagination: 'token', // ListObjectsV2's own ContinuationToken
  foreignKeys: false,
  canInsert: false,
  canUpdate: false,
  canDelete: false,
  writable: false,
  transactions: false,
  cancel: true, // the SDK's own abortSignal request option is fully effective, same as sqs/kafka
  // P33 commit 2 flips this to true alongside canInsert/canUpdate/canDelete/writable — false here
  // only for the brief window between this commit (the contract) and that one (the behaviour).
  fileTransfer: false,
};

File: apps/kira-studio/internal/adapters/s3/caps.go
package s3

import (
	"github.com/kirathecat/kira-studio/apps/kira-studio/internal/adapters"
	"github.com/kirathecat/kira-studio/apps/kira-studio/internal/page"
)

// caps is caps.ts's s3Caps, literally (P58d D18). Reuses the keyvalue shape — a single object's
// metadata+body is exactly a flat field/value listing, same as a redis hash.
var caps = adapters.Caps{
	Tabular:         false,
	Documents:       false,
	KeyValue:        true,
	Stream:          false,
	KeyBrowser:      true, // a bucket's prefix/object space is unbounded — browsed in a Browse tab
	DefaultPageKind: page.PageKindKeyValue,
	SQL:             false,
	// Stays false for now, as a named follow-up rather than a permanent no — an *object* already
	// shows its full metadata in the keyvalue view it opens into, so only a *bucket* has anything
	// new, and a bucket's properties are five separate SDK calls each of which a single-bucket IAM
	// policy routinely denies.
	Definition: false,
	// Describe throws E_UNSUPPORTED (adapter.go). Definition is already false above, so this is a
	// coincidence of two unrelated flags, not something Describe:false relies on.
	Describe:     false,
	Projection:   false,
	ServerFilter: false,
	// countObject (read.go) answers a single object's own field count via HeadObject, which is
	// always exact — the same per-item-exact resolution redis/caps.go makes for its own per-key
	// counts, not the bucket-wide "how many keys total" question ListObjectsV2 would need to
	// answer approximately.
	ExactCount:  true,
	Pagination:  adapters.PaginationToken, // ListObjectsV2's own ContinuationToken
	ForeignKeys: false,
	// An object's body is replaced wholesale via PutObject (CanUpdate), a new object is created
	// via a local-file upload (CanInsert), and DeleteObject removes one (CanDelete) — see
	// mutate.go. No per-object update in place; PutObject is always a full overwrite.
	CanInsert:    true,
	CanUpdate:    true,
	CanDelete:    true,
	Writable:     true,
	Transactions: false,
	Cancel:       true, // the op's own ctx aborting the in-flight SDK call is fully effective (P58d D3)
	// Gates Download outright and Upload together with CanInsert — the only engine whose items are
	// files with an OS dialog to move them through.
	FileTransfer: true,
}
